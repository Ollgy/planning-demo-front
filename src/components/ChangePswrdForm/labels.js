const labels = {
  oldPassword: 'Предыдущий пароль',
  newPassword: 'Новый пароль',
  rptPassword: 'Повторите новый пароль'
};

const placeholders = {
  oldPassword: '. . .',
  newPassword: '. . .',
  rptPassword: '. . .'
};

export { labels, placeholders };
