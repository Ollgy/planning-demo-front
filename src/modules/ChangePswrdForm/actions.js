import { createAction } from 'redux-actions';
import api from '../../api';

export const validateForm = createAction('VALIDATE_FORM_PSWRD');
export const inputForm = createAction('INPUT_FORM_PSWRD');
export const clearError = createAction('CLEAR_ERROR_PSWRD');

export const updatePasswordRequest = createAction('UPDATE_PASSWORD_REQUEST');
export const updatePasswordSuccess = createAction('UPDATE_PASSWORD_SUCCESS');
export const updatePasswordFailure = createAction('UPDATE_PASSWORD_FAILURE');

export const updatePassword = (userId, userData) => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(updatePasswordRequest());

      api
        .updateUser(userId, userData)
        .then(user => {
          dispatch(updatePasswordSuccess(user));
          resolve();
        })
        .catch(error => {
          dispatch(updatePasswordFailure(error));
          reject();
        });
    });
};
