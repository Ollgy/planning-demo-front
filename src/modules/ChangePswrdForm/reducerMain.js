import { combineReducers } from 'redux';
import passwordUpdate from './reducers/passwordUpdate';
import datalist from './reducers/datalist';

export default combineReducers({
  datalist,
  passwordUpdate
});
