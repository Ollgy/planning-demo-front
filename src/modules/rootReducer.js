import { combineReducers } from 'redux';
import authFromToken from './Auth';
import login from './Login';
import loginForm from './LoginForm';
import profile from './Profile';
import personalForm from './PersonalForm';
import changePswrdForm from './ChangePswrdForm';
import newTaskForm from './NewTaskForm';
import notes from './Notes';
import dashboard from './Dashboard';
import staff from './Staff';
import cleanBase from './CleanBase';

export default combineReducers({
  authFromToken,
  login,
  loginForm,
  profile,
  personalForm,
  changePswrdForm,
  newTaskForm,
  notes,
  dashboard,
  staff,
  cleanBase
});
