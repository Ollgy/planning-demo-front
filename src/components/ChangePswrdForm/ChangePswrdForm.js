import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import Button from '../Button';
import Input from '../Input';
import FormMsg from '../FormMsg';
import CloseButton from '../CloseButton';
import { labels, placeholders } from './labels';
import {
  inputForm,
  validateForm,
  clearError,
  updatePassword
} from '../../modules/ChangePswrdForm';
import styles from '../Profile/Profile.module.css';
import formStyles from '../LoginForm/LoginForm.module.css';

class ChangePswrdForm extends PureComponent {
  constructor(props) {
    super(props);
    this.activeInput = React.createRef();
  }

  componentDidMount() {
    this.activeInput.current.focus();
  }

  onInputChange = e => {
    const { inputForm, clearError } = this.props;
    const { value, name } = e.target;

    inputForm({ input: name, value });
    clearError({ input: name });
  };

  onSubmit = async e => {
    e.preventDefault();

    const {
      user,
      fields,
      updatePassword,
      validateForm,
      hidePswrdBlock
    } = this.props;

    await validateForm(fields);

    const { validation } = this.props;

    if (validation.status) {
      await updatePassword(user.id, {
        oldPassword: fields.oldPassword,
        password: fields.newPassword
      });

      hidePswrdBlock();
    } else {
      this.activeInput.current.focus();
    }
  };

  render() {
    const { fields, validation, hidePswrdBlock } = this.props;
    const { msg, error, errorInputs } = validation;

    return (
      <form
        action="/"
        className={styles.formChangePswrd}
        onSubmit={this.onSubmit}
      >
        <div className={styles.closeButtonBlock}>
          <CloseButton onClick={hidePswrdBlock} size="m" />
        </div>
        <div className={styles.infoBlock}>
          {Object.keys(labels).map((field, i) => (
            <div className={formStyles.formItem} key={field}>
              <Input
                type="password"
                ref={
                  errorInputs.length
                    ? field === errorInputs[0]
                      ? this.activeInput
                      : null
                    : i === 0
                    ? this.activeInput
                    : null
                }
                errorClass={errorInputs.includes(field) ? error : null}
                name={field}
                value={fields[field]}
                placeholder={placeholders[field]}
                onChange={this.onInputChange}
              />
              <label className={formStyles.formLabel} htmlFor={field}>
                <span>{labels[field]}</span>
              </label>
            </div>
          ))}
        </div>
        <div className={styles.msgBlock}>
          <FormMsg children={msg} />
        </div>
        <div className={styles.buttonBlock}>
          <Button type="submit" name="submit" children="Изменить" />
        </div>
      </form>
    );
  }
}

const mapStateToProps = state => ({
  fields: state.changePswrdForm.datalist.fields,
  validation: state.changePswrdForm.datalist.validation,
  passwordUpdate: {
    user: state.changePswrdForm.passwordUpdate.user,
    isLoading: state.changePswrdForm.passwordUpdate.isLoading,
    error: state.changePswrdForm.passwordUpdate.error
  }
});

const mapDispatchToProps = dispatch => ({
  updatePassword: (arg1, arg2) => dispatch(updatePassword(arg1, arg2)),
  inputForm: arg => dispatch(inputForm(arg)),
  validateForm: arg => dispatch(validateForm(arg)),
  clearError: arg => dispatch(clearError(arg))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ChangePswrdForm);
