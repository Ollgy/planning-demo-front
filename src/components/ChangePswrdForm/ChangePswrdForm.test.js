import React from 'react';
import ChangePswrdForm from './ChangePswrdForm';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { createStore, compose, applyMiddleware } from 'redux';
import rootReducer from '../../modules/rootReducer';
import thunk from 'redux-thunk';

const store = createStore(rootReducer, {}, compose(applyMiddleware(thunk)));

describe('<ChangePswrdForm/> UI component', () => {
  const shallowExpect = compose(
    expect,
    toJSON,
    shallow
  );

  it('Renders correctly', () => {
    shallowExpect(<ChangePswrdForm store={store} />).toMatchSnapshot();
  });

  const wrapper = shallow(<ChangePswrdForm store={store} />);
  const inner = wrapper.children().first();

  it('contains fields in props', () => {
    expect(inner.props().fields).toBeTruthy();
  });
});
