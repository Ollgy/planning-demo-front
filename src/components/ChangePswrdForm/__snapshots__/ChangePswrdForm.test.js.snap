// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangePswrdForm/> UI component Renders correctly 1`] = `
<ContextProvider
  value={null}
>
  <ChangePswrdForm
    clearError={[Function]}
    fields={
      Object {
        "newPassword": "",
        "oldPassword": "",
        "rptPassword": "",
      }
    }
    inputForm={[Function]}
    passwordUpdate={
      Object {
        "error": null,
        "isLoading": false,
        "user": null,
      }
    }
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    updatePassword={[Function]}
    validateForm={[Function]}
    validation={
      Object {
        "error": "",
        "errorInputs": Array [],
        "msg": "",
        "status": false,
      }
    }
  />
</ContextProvider>
`;

exports[`<PersonalForm/> UI component Renders correctly 1`] = `
<ContextProvider
  value={null}
>
  <ChangePswrdForm
    clearError={[Function]}
    fields={
      Object {
        "newPassword": "",
        "oldPassword": "",
        "rptPassword": "",
      }
    }
    inputForm={[Function]}
    passwordUpdate={
      Object {
        "error": null,
        "isLoading": false,
        "user": null,
      }
    }
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    updatePassword={[Function]}
    validateForm={[Function]}
    validation={
      Object {
        "error": "",
        "errorInputs": Array [],
        "msg": "",
        "status": false,
      }
    }
  />
</ContextProvider>
`;
